### ISSUE TITLE
leak in backendBandwidth() and backendChunks() filters

### ISSUE THREAD

**Issue by aryszka** [zalando/skipper#282]
the change to the handling of the outgoing request body in:

https://github.com/zalando/skipper/pull/277

...causes a leak in the following filters in the diag package:

- backendBandwidth()
- backendChunks()

the bug because the request body is expected to be closed by the RoundTripper. These filters spawn a goroutine and set an io.PipeReader as the body. This goroutine never terminates unless the body is closed, which happens when the incoming request doesn't have a body. The mechanism in this case looks like this:

1. the incoming server side request has a body object, even if the request has no content

```
    // For server requests the Request Body is always non-nil
    // but will return EOF immediately when no body is present.
    // The Server will close the request body. The ServeHTTP
    // Handler does not need to.
```

2. the backendChunks filter will create an io.Pipe, set replacing the body with the pipe reader, and spawning a goroutine that copies the original body to the pipe writer, which proceeds only when the pipe reader is read

3. the proxy drops the body (the pipe reader), and never closes it

I see currently two potential solutions:

A: fix this in the filters, currently only the above mentioned two filters. This means that the filter would need to set the ContentLength to either < 0 or > 0, if it sets the body. The drawback is that this is not necessarily obvious from the net/http docs that this should be done.

B: in the proxy, check if the body object was changed to something else during the request filter chain

opinions, insights, suggestions?

**Comment by lmineiro** [comment]
I believe B is the least bad approach.
That being said, I also think filters need to act responsibly. The example given could be considered one of those cases. There are many other ways filters can break things and only so much the engine itself can do to prevent them from misbehaving.

**Comment by aryszka** [comment]
good point. so the plan now is to apply B, but also make the existing filters behave. Also point it out in the docs.
